sql/binlog.cc

```diff
@@ -8167,6 +8168,13 @@ void MYSQL_BIN_LOG::process_after_commit_stage_queue(THD *thd, THD *first) {
       Thd_backup_and_restore switch_thd(thd, head);
       bool all = head->get_transaction()->m_flags.real_commit;
       (void)RUN_HOOK(transaction, after_commit, (head, all));
+
+      if (rpl_wait_for_semi_sync_ack) {
```

Comment:
If signal_semi_sync_ack doesn't get called on every process_after_commit, I think there's a potential problem here. If no further writes come into the system, but rpl_wait_for_semi_sync_ack is enabled, wouldn't async slaves be held back at the last_acked checkpoint that is much much older than the latest semi-sync acked checkpoint?

It seems like we should update the last_acked position for each commit.

It also seems like the original code might have this issue too.

Reply:
@hermanlee I'm not sure i understand this statement: " last_acked checkpoint that is much much older than the latest semi-sync acked checkpoint"

Reply:
I think the scenario is this:

rpl_wait_for_semi_sync_ack is disabled
master processes writes and generates 1GB worth of binlog entries
update binlog entry is not called, so the async slave's last_acked checkpoint occurs before the new binlog entries
writes stop
rpl_wait_for_semi_sync_ack is enabled
async slaves connects, but last_acked checkpoint before the latest 1GB of binlog entries

Reply:
I think that case is handled in update_rpl_wait_for_semi_sync_ack() ?

Reply:
The only place I see last_acked updated is in signal_semi_sync_ack(). How does update_rpl_wait_for_semi_sync_ack() trigger signal_semi_sync_ack()? The only calls I see to signal_semi_sync_ack() is through process_after_commit_stage_queue(), which only happens when writes come into the system.

Maybe update_rpl_wait_for_semi_sync_ack() should update last_ack somehow too?

Reply:
Won't `Binlog_sender::get_binlog_end_pos()` be called after `mysql_bin_log.signal_update()` is called in `update_rpl_wait_for_semi_sync_ack()`? If yes, then this call to `Binlog_sender::get_binlog_end_pos()` will now call `mysql_bin_log.get_binlog_end_pos()` with `async_slave = false` so it'll return the latest pos instead of the last acked pos after rpl_wait_for_semi_sync_ack is disabled. I don't think we need to update `last_acked` but I might be missing something.

Reply:
I think I see where the disconnect is. The question is what should be the expected behavior when we transition from rpl_wait_for_semi_sync from on to off to on.

Here's the case: rpl_wait_for_semi_sync is on, semi-sync ackers are running, last_ack is updated, so all is working as intended.

Now, rpl_wait_for_semi_sync is disabled, semi_sync ackers are removed. Async slaves connected to the server can tail all binlogs.

The master server generates 1GB of binlog entries after rpl_wait_for_semi_sync is disabled. I think we agree last_ack is pointing to the binlog entry before the latest 1GB was created (assuming they are all in the same binlog file).

Now someone re-enables rpl_wait_for_semi_sync.

Is the expectation for async slaves that connect now to be waiting for the last 1GB of binlogs to be acked by a semi-sync acker before they can read them? From my understanding, last_ack is only updated when semi-sync ackers respond. Or is the expectation for async slaves to be able to read binlogs up to when rpl_wait_for_semi_sync is enabled, which was my expectation, but it might not be what is desired in all scenarios.

The code seems to support the former, but I was thinking of the latter.

Reply:
I see you concern. In the original FB patch this case was handled by updating the last acked position on initial semi-sync slave connection (it was updated to the point where the semi-sync slave is asking binlogs from because we can assume that all trxs were acked until that point).

Reply:
It sounds like we'd want the same behavior? Keep last_acked as is until a new semi-sync slave starts asking for binlogs? Update last_ack to that position if needed? If there are no semi-sync ackers, then it sounds like the original was to keep last_ack untouched and block async slaves from anything that hasn't been acked by a semi-sync acker.

@abhinav04sharma Is this correct?

Reply:
That is correct.

Reply:
I did a lot of research and in 8.0 it's not possible to use information from initial semi-sync slave connection to update `last_ack`. With `gtid-mode=ON` the initial connection sends only an executed GTID set. I think that the best solution would be to remove `if (rpl_wait_for_semi_sync_ack)` and update `last_ack` unconditionally.

Reply:
@inikep even if we remove the rpl_wait_for_semi_sync_ack check the signalling will only happen when something is written to the binlog, correct? So will the async slaves not wait until something is new is written to the binlog (in @hermanlee example above) ?

Reply:
Nevermind, looks like last_acked in maintained in mysql_bin_log which is a global singleton, so this should work. I'll let @hermanlee comment on the proposed solution.

Reply:
I have a new idea. What do you think about calling `signal_semi_sync_ack()` only when `semisync_master.so` is loaded (installed)? It can be easily done using `plugin_dl_find()` from `sql_plugin.cc`.

Reply:
I think updating unconditionally would be fine.

@abhinav04sharma is the recent problem found with ABS an issue with this port to 8.0?

Reply:
That issue was around group commit, it should not be a problem in this patch since we're signalling inside `ordered_commit()`.

Regarding when to call `signal_semi_sync_ack()`, i think even we always call it (i.e. even if rpl_wait_for_semi_sync_ack = 0) we'll still have the same problem if the master restarts since the last_acked will be reset and we'll need at least one write to start sending events to lagging async slaves... Not sure how we can solve this.